org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/StringBasedTextRegionAccessDiffAppender.java

```diff
@@ -52,132 +58,149 @@ protected StringHiddenRegion appendHiddenRegion(boolean undefined) {
+		StringHiddenRegion hidden = (StringHiddenRegion) this.last;
 		EObject semanticElement = source.getSemanticElement();
 		AbstractEObjectRegion region = getOrCreateEObjectRegion(semanticElement, source.getEObjectRegion());
-		int offset = access.append(text);
+		String changedText = textChanges.get(source);
+		String newText = changedText != null ? changedText : source.getText();
+		int offset = result.append(newText);
 		EObject grammar = source.getGrammarElement();
-		StringHiddenRegion previous = (StringHiddenRegion) this.last;
-		StringSemanticRegion result = new StringSemanticRegion(access, region, grammar, offset, text.length());
+		StringSemanticRegion result = new StringSemanticRegion(this.result, region, grammar, offset, newText.length());
 		region.addChild(result);
-		previous.setNext(result);
-		result.setLeadingHiddenRegion(previous);
+		hidden.setNext(result);
+		result.setLeadingHiddenRegion(hidden);
 		this.last = result;
+		recordDiff(source, result);
+		if (changedText != null && this.diffNesting == 0) {
+			this.result.append(new SequentialRegionDiff(source, source, result, result));
+		}
 		return result;
 	}
 
-	protected ISequentialRegion copyAndAppend(ISequentialRegion first, ISequentialRegion last) {
-		ISequentialRegion current = first;
-		ISequentialRegion result = null;
+	public void copyAndAppend(ISemanticRegion substituteFirst, ISemanticRegion substituteLast) {
+		if (!(this.last instanceof StringHiddenRegion)) {
+			appendHiddenRegion(true);
+		}
+		ISequentialRegion current = substituteFirst;
 		while (true) {
 			if (current instanceof IHiddenRegion) {
 				copyAndAppend((IHiddenRegion) current);
 			} else if (current instanceof ISemanticRegion) {
-				copyAndAppend((ISemanticRegion) current, current.getText());
+				copyAndAppend((ISemanticRegion) current);
 			}
-			if (result == null) {
-				result = this.last;
-			}
-			if (current == last) {
+			if (current == substituteLast) {
 				break;
 			}
 			current = current.getNextSequentialRegion();
 			if (current == null) {
 				throw new IllegalStateException("last didn't match");
 			}
 		}
-		return result;
 	}
 
-	protected void copyHiddenRegionParts(StringHiddenRegion region, Iterable<IHiddenRegionPart> parts) {
-		for (IHiddenRegionPart part : parts) {
-			String text = part.getText();
-			int offset = access.append(text);
-			if (part instanceof IComment) {
-				IComment comment = ((IComment) part);
-				AbstractRule grammarElement = (AbstractRule) comment.getGrammarElement();
-				StringComment newComment = new StringComment(region, grammarElement, offset, text.length());
-				region.addPart(newComment);
-			} else if (part instanceof IWhitespace) {
-				IWhitespace ws = (IWhitespace) part;
-				AbstractRule grammarElement = (AbstractRule) ws.getGrammarElement();
-				StringWhitespace newWs = new StringWhitespace(region, grammarElement, offset, text.length());
-				region.addPart(newWs);
-			}
-		}
-	}
+	private boolean diffDelete = false;
```

Comment:
Could you please group the fields?

Reply:
see: https://github.com/eclipse/xtext-core/pull/652